File: back/API/Routes/customers/getCustomers.py
from base64 import b64encode
from flask import Blueprint, jsonify
from dbConnection import db
from ...JWT_manager import jwt
from flask_jwt_extended import jwt_required
from ...decorators import role_required
from gridfs import GridFS

fs = GridFS(db)

get_customers_blueprint = Blueprint('get_customers', __name__)

@get_customers_blueprint.route('/api/customers/<customer_id>', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomer(customer_id):
    customer = db.customers.find_one(
        {'id': int(customer_id)},
        {'_id': 0, 'paymentsHistory': 0, 'clothes': 0, 'saved_clothes': 0, 'encounters': 0, 'image': 0}
    )
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404

    return jsonify(customer), 200


@get_customers_blueprint.route('/api/customers', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomers():
    customers = db.customers.find(
        {},
        {'_id': 0, 'paymentsHistory': 0, 'clothes': 0, 'saved_clothes': 0, 'encounters': 0, 'image': 0}
    )
    if customers is None:
        return jsonify({'details': 'No customers found'}), 404

    return jsonify(list(customers)), 200


@get_customers_blueprint.route('/api/customers/<customer_id>/image', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomerImage(customer_id):
    customer = db.customers.find_one({'id': int(customer_id)}, {'_id': 0, 'image': 1})
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404

    image_data = fs.get(customer['image']).read()
    base64_image = b64encode(image_data).decode('utf-8')
    if image_data is None:
        return jsonify({'details': 'Image not found'}), 404

    return jsonify({ 'image': base64_image }), 200


@get_customers_blueprint.route('/api/customers/<customer_id>/payments_history/<payment_id>', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomerPayment(customer_id, payment_id):
    customer = db.customers.find_one({'id': int(customer_id)})
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404

    payment = db.customers.find_one(
        {'id': int(customer_id), 'payments_history.id': int(payment_id)},
        {'_id': 0, 'payments_history.$': 1}
    )
    if payment is None:
        return jsonify({'details': 'Payment not found'}), 404

    return jsonify(payment), 200


@get_customers_blueprint.route('/api/customers/<customer_id>/payments_history', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomerPayments(customer_id):
    customer = db.customers.find_one({'id': int(customer_id)})
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404

    return jsonify(customer.get('payments_history', [])), 200


@get_customers_blueprint.route('/api/customers/<customer_id>/clothes/<clothe_id>', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomerClothe(customer_id, clothe_id):
    customer = db.customers.find_one({'id': int(customer_id)})
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404

    clothe = db.customers.find_one(
        {'id': int(customer_id), 'clothes.id': int(clothe_id)},
        {'_id': 0, 'clothes.$': 1}
    )
    if clothe is None:
        return jsonify({'details': 'Clothe not found'}), 404

    image_data = fs.get(customer['image']).read()
    base64_image = b64encode(image_data).decode('utf-8')
    return jsonify({
        'id': clothe['id'],
        'type': clothe['type'],
        'image': base64_image
    }), 200


@get_customers_blueprint.route('/api/customers/<customer_id>/clothes', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomerClothes(customer_id):
    customer = db.customers.find_one({'id': int(customer_id)})
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404
    clothes_with_images = []
    for clothe in customer['clothes']:
        image_data = GridFS(db).get(clothe['image']).read()
        base64_image = b64encode(image_data).decode('utf-8')
        clothes_with_images.append({
            'id': clothe['id'],
            'type': clothe['type'],
            'image': base64_image
        })
    return jsonify(clothes_with_images)


@get_customers_blueprint.route('/api/customers/<customer_id>/saved_clothes', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomerSavedClothes(customer_id):
    customer = db.customers.find_one({'id': int(customer_id)})
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404
    clothes_with_images = []
    for clothe in customer['saved_clothes']:
        image_data = GridFS(db).get(clothe['image']).read()
        base64_image = b64encode(image_data).decode('utf-8')
        clothes_with_images.append({
            'id': clothe['id'],
            'type': clothe['type'],
            'image': base64_image
        })
    return jsonify(clothes_with_images)


@get_customers_blueprint.route('/api/customers/<customer_id>/encounters/<encounter_id>', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomerEncounter(customer_id, encounter_id):
    customer = db.customers.find_one({'id': int(customer_id)})
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404

    encounter = db.customers.find_one(
        {'id': int(customer_id), 'encounters.id': int(encounter_id)},
        {'_id': 0, 'encounters.$': 1}
    )
    if encounter is None:
        return jsonify({'details': 'Encounter not found'}), 404

    return jsonify(encounter), 200


@get_customers_blueprint.route('/api/customers/<customer_id>/encounters', methods=['GET'])
@jwt_required()
@role_required('Coach')
def getCustomerEncounters(customer_id):
    customer = db.customers.find_one({'id': int(customer_id)})
    if customer is None:
        return jsonify({'details': 'Customer not found'}), 404

    return jsonify(customer.get('encounters', [])), 200

File: front/src/app/services/customers/get-customers.ts
import { Clothe, isClothe, isClothes } from "@/app/types/Clothe"
import { BasicCustomerWithID, Customer, isCustomer, isCustomers } from "@/app/types/Customer"
import { Encounter, isEncounter, isEncounters } from "@/app/types/Encounter"
import { Payment, isPayment, isPaymentsHistory } from "@/app/types/PaymentHistory"
import FetchError from '@/app/types/FetchErrors'
import CustomersService from '@/app/services/customers/class-customers'

export default class GetCustomersService extends CustomersService {
  public async getCustomer(data: { id: number }): Promise<BasicCustomerWithID> {
    const res = await fetch(`${this._route}/${data.id}`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!isCustomer(object))
      throw new Error(`An error happened when fetching customer ${data.id}.`, { cause: `Returned object doesn't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }

  public async getCustomers(): Promise<Customer[]> {
    const res = await fetch(this._route, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching customers.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!isCustomers(object))
      throw new Error('An error happened when fetching customers.', { cause: `Returned objects don't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }

  public async getCustomerImage(data: { id: number }): Promise<string> {
    const res = await fetch(`${this._route}/${data.id}/image`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer image.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (typeof object.image !== 'string')
      throw Error(`An error happened when fetching the customer ${data.id} image.`, { cause: `Returned object doesn't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object.image
  }

  public async getCustomerPayment(data: { id: number, payment_id: number }): Promise<Payment> {
    const res = await fetch(`${this._route}/${data.id}/payments_history/${data.payment_id}`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer payment.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!isPayment(object))
      throw new Error(`An error happened when fetching customer ${data.id} payment ${data.payment_id}.`, { cause: `Returned object doesn't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }

  public async getCustomerPaymentsHistory(data: { id: number }): Promise<Payment[]> {
    const res = await fetch(`${this._route}/${data.id}/payments_history`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer payments history.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!isPaymentsHistory(object))
      throw new Error(`An error happened when fetching customer ${data.id} payments history.`, { cause: `Returned objects don't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }

  public async getCustomerClothe(data: { id: number, clothe_id: number }): Promise<Clothe> {
    const res = await fetch(`${this._route}/${data.id}/clothes/${data.clothe_id}`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer clothe.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!isClothe(object))
      throw new Error(`An error happened when fetching customer ${data.id} clothe ${data.clothe_id}.`, { cause: `Returned object doesn't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }

  public async getCustomerClothes(data: { id: number }): Promise<Clothe[]> {
    const res = await fetch(`${this._route}/${data.id}/clothes`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer clothes.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!isClothes(object))
      throw new Error(`An error happened when fetching customer ${data.id} clothes list.`, { cause: `Returned objects don't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }

  public async getCustomerSavedClothes(data: { id: number }): Promise<number[]> {
    const res = await fetch(`${this._route}/${data.id}/saved_clothes`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer saved clothes.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!Array.isArray(object))
      throw new Error(`An error happened when fetching customer ${data.id} saved clothes list.`, { cause: `Returned objects don't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }

  public async getCustomerEncounter(data: { id: number, encounter_id: number }): Promise<Encounter> {
    const res = await fetch(`${this._route}/${data.id}/encounters/${data.encounter_id}`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer encounter.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!isEncounter(object))
      throw new Error(`An error happened when fetching customer ${data.id} encounter ${data.encounter_id}.`, { cause: `Returned object doesn't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }

  public async getCustomerEncounters(data: { id: number }): Promise<Encounter[]> {
    const res = await fetch(`${this._route}/${data.id}/encounters`, {
      method: 'GET',
      headers: this._headers,
      credentials: 'include'
    })
    const object = await res.json()

    if (!res.ok) {
      throw new FetchError({
        message: 'An error occured when fetching a customer encounters.',
        status: res.status,
        statusText: res.statusText,
        details: object.details
      })
    }

    if (!isEncounters(object))
      throw new Error(`An error happened when fetching customer ${data.id} encounters list.`, { cause: `Returned objects don't correspond to the associated type.\n${JSON.stringify(object)}` })

    return object
  }
}
